### ISSUE TITLE
Don't use GetPrivateProfile* WinAPI functions for loading INI

### ISSUE THREAD

**Issue by phobos2077** [sfall-team/sfall#468]
This is rather big issue considering how much INI option sfall itself loads, not to mention mods (especially EcCo, has a lot of ini settings, and will have even more...).

For every call to read a parameter (from sfall or script):

- It searches Windows Registry
- IF it can't find anything (and it won't ofc), then it loads and parses the entire ini file
- Then, returns a requested value

So instead we need an Ini Parser that:

- On first request, parses the ini file and saves sections & key/value pairs in dictionaries
- Saves the parsed INI memory afterwards
- If INI was requested from a script, it should unload it from memory on game reload (because we reset everything at that point, this is how we iterate on mods now).
- For sfall settings, it can keep them loaded at all times (we don't expect changes to INI to take effect without game restart)
- It should support loading ini from VFS (dat files), need to think about some kind of path prefix, or another solution so that root-based and mods-folder-based paths still work.

One option is to use built-in Fallout Config parser  (config_init, config_load, config_get_value, etc). Need to figure out how to get a list of sections and list of keys in section from that. config_load has parameter that allows to choose to load INI from "db" or from disk directly, which should cover the last requirement.

Another option is to write our own parser (or "adopt" some else's).

**Comment by phobos2077** [comment]
Some analysis of FO2 built in parser:

- It stores data in a 2-level sorted dictionary (first level is section name => section, sections are key=>value pairs).
- Dictionaries use half division method to find values by key, which is O(log(n))
- When adding new key/value pair, they first find an index to place they key based on order, then shift elements from the end to the new index - O(n)
- Config parser supports comments after section header as well as value (GetPrivateProfile** doesn't - it returns strings with the comment, which may break some existing script if for some dumb reason it relies on this "feature")
- Because it reads the file (txt or cfg) once, it should be much more efficient than current Sfall functions, even without the cost of reading from registry. Just by the fact that we save on disk IO operations.

So I suggest for ddraw.ini definitely only read Config once before init (and maybe dispose it after). For INI requests from scripts we should provide our own `unordered_map<string, Config>`. If key not found, then use FO2 parser to parse the ini. Important to unload configs and clear this map on game reload (to be able to quickly iterate mod settings).

**Comment by phobos2077** [comment]
In all this excitement I forgot that scripts can write INI, not just read from it. This complicates matters somewhat, as FO parser can't write. And if you use parser like this to write, you will loose all of the comments. So for writing we have to:

1) Use separate INI parser/writer only for writing that would be able to open file, find line in it (if it exists), and then write it, keeping comments intact if possible.
2) Keep using WritePrivateProfileString function that already does that. Doesn't matter if it's inefficient as much, since current mods don't do INI writes often. But who knows, this may change.

**Comment by phobos2077** [comment]
Using config parser from the engine directly didn't work well. I was able to read data no problem, but I couldn't simply cache these configs (keep them for whole game session and call `config_exit` on reload). There was a consistent memory access violation for some reason. I could just replace all memory allocation calls within `config_*` and `dictionary*` functions, but opted instead to re-implement these functions partially in normal C++.

Seems to work perfectly so far for all ini reads, sfall and script. For "set" functions, I went for option (2) from the last comment.
